File: src/main/java/com/polygloat/model/ApiKey.java
package com.polygloat.model;

import com.polygloat.constants.ApiScope;
import lombok.*;

import javax.persistence.*;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import java.util.Arrays;
import java.util.Set;
import java.util.stream.Collectors;

@Entity
@Table(uniqueConstraints = {
        @UniqueConstraint(columnNames = {"key"}, name = "api_key_unique"),
})
@EqualsAndHashCode(callSuper = true, of = {"id", "key", "scopes"})
@NoArgsConstructor
@AllArgsConstructor
@ToString(of = {"key"})
@Builder
@Data
public class ApiKey extends AuditModel {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotNull
    @ManyToOne
    private UserAccount userAccount;

    @NotNull
    @ManyToOne
    private Repository repository;

    @NotEmpty
    @NotNull
    private String key;

    @NotBlank
    @NotNull
    private String scopes;

    public Set<ApiScope> getScopes() {
        return Arrays.stream(this.scopes.split(";")).map(ApiScope::fromValue).collect(Collectors.toSet());
    }

    public void setScopes(Set<ApiScope> scopes) {
        this.scopes = stringifyScopes(scopes);
    }

    @SuppressWarnings("unused")
    public static class ApiKeyBuilder {
        private String scopes;

        public ApiKeyBuilder scopes(Set<ApiScope> scopes) {
            this.scopes = stringifyScopes(scopes);
            return this;
        }
    }

    private static String stringifyScopes(Set<ApiScope> scopes) {
        return scopes.stream().map(ApiScope::getValue).collect(Collectors.joining(";"));
    }
}

File: src/main/kotlin/io/tolgee/model/ApiKey.kt
package io.tolgee.model

import io.tolgee.constants.ApiScope
import javax.persistence.*
import javax.validation.constraints.NotEmpty
import javax.validation.constraints.NotNull

@Entity
@Table(uniqueConstraints = [UniqueConstraint(columnNames = ["key"], name = "api_key_unique")])
data class ApiKey(
        @Id
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        var id: Long? = null,

        @NotEmpty
        @NotNull
        var key: String? = null,

        @NotNull
        @NotEmpty
        @Enumerated(EnumType.ORDINAL)
        @field:ElementCollection(targetClass = ApiScope::class, fetch = FetchType.EAGER)
        var scopesEnum: Set<ApiScope>
) : AuditModel() {

    @ManyToOne
    @NotNull
    var userAccount: UserAccount? = null

    @ManyToOne
    @NotNull
    var project: Project? = null

    @Deprecated(level = DeprecationLevel.WARNING,
            message = "Scopes field is deprecated, it should not persist string values",
            replaceWith = ReplaceWith("scopesEnum")
    )
    private var scopes: String? = null

    constructor(
            id: Long? = null,

            @NotNull
            userAccount: UserAccount?,

            @NotNull
            project: Project?,

            @NotEmpty @NotNull
            key: String?,

            scopes: String? = null,

            @NotEmpty @NotNull
            scopesEnum: Set<ApiScope>
    ) : this(id, key, scopesEnum) {
        this.userAccount = userAccount
        this.project = project
        @Suppress("DEPRECATION")
        this.scopes = scopes
    }
}
